include/zephyr/math/q_ops.h

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2022 Google Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef INCLUDE_ZEPHYR_MATH_Q_OPS_H_
+#define INCLUDE_ZEPHYR_MATH_Q_OPS_H_
+
+#include <zephyr/math/q_types.h>
+#include <zephyr/sys/__assert.h>
+
+#ifndef INCLUDE_MATH_FP_H_
+#error "Please include <zephyr/math/fp.h> instead"
+#endif
+
+/* clang-format off */
+#define q_assign(qval, x)                                                                          \
+	do {                                                                                       \
+		int64_t intermediate =                                                             \
+			(int64_t) _Generic(                                                        \
+				(x),                                                               \
+				uint8_t : ((int64_t)x << 32),                                      \
+				uint16_t : ((int64_t)x << 32),                                     \
+				uint32_t : ((int64_t)x << 32),                                     \
+				int8_t : ((int64_t)x << 32),                                       \
+				int16_t : ((int64_t)x << 32),                                      \
+				int32_t : ((int64_t)x << 32),                                      \
+				float : ((float)x * (float)(INT64_C(1) << 32)),                    \
+				double : ((float)x * (double)(INT64_C(1) << 32)),                  \
+				q1_8_t : ((int64_t)x << 24),                                       \
+				q1_16_t : ((int64_t)x << 16),                                      \
+				q1_31_t : ((int64_t)x << 1));                                      \
```

Comment:
@teburd I can't get this part to work. Do you have any ideas? The issue is that `_Generic` needs to compile for every branch of the statements and I can't cast `q1_8_t` to `int64_t`. Alternatively, if I were to use `x.value` this will fail if I pass `INT32_C(5)` for `x`.

Reply:
Got it working by using a function pointer